p2p/conn/connection_test.go

```diff
@@ -769,3 +770,147 @@ func stopAll(t *testing.T, stoppers ...stopper) func() {
 		}
 	}
 }
+
+func fib(n int) int {
+	if n < 2 {
+		return n
+	}
+	return fib(n-1) + fib(n-2)
+}
+
+// GenerateMessages generates messages of a given size at specified rate `messagingRate`
+// for a given duration `totalDuration` until the total number of messages
+// `totalNum` is reached. If `totalNum` is less than zero,
+// then the message generation continues until the `totalDuration` is reached.
+func GenerateMessages(mc *MConnection, messagingRate time.Duration,
+	totalDuration time.Duration, totalNum int, msgSize int, chID byte) {
+	// all messages have an identical content
+	msg := bytes.Repeat([]byte{'x'}, msgSize)
+
+	// message generation interval ticker
+	ticker := time.NewTicker(messagingRate)
+	defer ticker.Stop()
+
+	// timer for the total duration
+	timer := time.NewTimer(totalDuration)
+	defer timer.Stop()
+
+	sentNum := 0
+	// generating messages
+	for {
+		select {
+		case <-ticker.C:
+			// generate message
+			if mc.Send(chID, msg) {
+				sentNum++
+				if totalNum > 0 && sentNum >= totalNum {
+					log.TestingLogger().Info("Completed the message generation as the" +
+						" total number of messages is reached")
+					return
+				}
+			}
+		case <-timer.C:
+			// time's up
+			log.TestingLogger().Info("Completed the message generation as the total " + "duration is reached")
+			return
+		}
+	}
+}
+
+func BenchmarkMConnection(b *testing.B) {
+	msgSize := 5 * 1024 // in bytes
+	totalMsg := 100     // total number of messages to be sent
+	chID := byte(0x01)
+	SendQueueCapacity := 10 // in messages
+
+	b.Run("benchmark MConnection", func(b *testing.B) {
+		//cpuFile, _ := os.Create("cpu.pprof")
+		//pprof.StartCPUProfile(cpuFile)
+		//defer pprof.StopCPUProfile()
+
+		//f, _ := os.Create("block.pprof")
+		//runtime.SetBlockProfileRate(1)
+
+		for n := 0; n < b.N; n++ {
+			//	set up two networked connections
+			//server, client := NetPipe()
```

Comment:
Most tests for MConnection utilize NetPipe() which implements an in-memory, unbuffered connections. To avoid potential performance bottlenecks in the benchmark and to more accurately reflect MConnection's behavior in a production environment, these have been substituted with tcp connections on localhost.